File: RepoTransBench/target_projects/Python/C++/scrapy_cssselect/tests/original/test_xpath_expr.cpp
#include <gtest/gtest.h>
#include <string>

// Dummy XPathExpr Class to match used API in tests (replace with real implementation)
class XPathExpr {
public:
    std::string path;
    std::string element;
    std::string condition;

    XPathExpr(const std::string& p, const std::string& el, const std::string& cond = "")
        : path(p), element(el), condition(cond) {}

    std::string str() const {
        if (!condition.empty())
            return path + element + "[" + condition + "]";
        return path + element;
    }

    void add_condition(const std::string& cond) {
        if (!condition.empty())
            condition += " and ";
        condition += cond;
    }

    void add_name_test() {
        element = "*";
        if (!condition.empty()) {
            condition += " and name()";
        } else {
            condition = "name()";
        }
    }

    void add_star_prefix() {
        // Simulate adding star prefix by changing path to include '*'
        if (path.back() != '*')
            path += "*";
    }

    XPathExpr join(const std::string& combiner, const XPathExpr& other, const std::string& closing_combiner, bool has_inner_condition) const {
        XPathExpr result(
            path + combiner + other.path + closing_combiner,
            other.element,
            other.condition
        );
        return result;
    }
};

TEST(TestXPathExpr, StrAndAddCondition) {
    XPathExpr x("//", "div", "foo=1");
    ASSERT_EQ(x.str(), "//div[foo=1]");
    x.add_condition("bar=2");
    ASSERT_EQ(x.condition.find("bar=2") != std::string::npos, true);
    ASSERT_EQ(x.condition.find("[foo=1)") == std::string::npos, true);
}

TEST(TestXPathExpr, AddNameTest) {
    XPathExpr x("//", "div");
    x.add_name_test();
    ASSERT_EQ(x.element, "*");
    ASSERT_TRUE(x.condition.find("name()") != std::string::npos);
}

TEST(TestXPathExpr, AddStarPrefix) {
    XPathExpr x("//", "*", "");
    x.add_star_prefix();
    ASSERT_TRUE(x.path == "//*" || x.path.back() == '*');
}

TEST(TestXPathExpr, Join) {
    XPathExpr x1("//", "a", "foo=1");
    XPathExpr x2("/*/", "span", "bar=2");
    XPathExpr r = x1.join("|", x2, "::", true);
    ASSERT_TRUE(r.element.find("span") == 0);
    ASSERT_TRUE(r.path.find("|") != std::string::npos);
}

File: RepoTransBench/target_projects/Python/Rust/scrapy_cssselect/tests/original/test_xpath_expr.rs
use scrapy_cssselect_rs::xpath::XPathExpr;

#[test]
fn test_xpath_expr_str_and_add_condition() {
    let mut x = XPathExpr::new("//", "div", "foo=1");
    assert_eq!(x.to_string(), "//div[foo=1]");
    x.add_condition("bar=2");
    // Check that unbalanced parens are not present and bar=2 got added
    assert!(!x.condition.contains("[foo=1)"));
    assert!(x.condition.contains("bar=2"));
}

#[test]
fn test_xpath_expr_add_name_test() {
    let mut x = XPathExpr::new("//", "div", "");
    x.add_name_test();
    assert_eq!(x.element, "*");
    assert!(x.condition.contains("name()"));
}

#[test]
fn test_xpath_expr_add_star_prefix() {
    let mut x = XPathExpr::new("//", "*", "");
    x.add_star_prefix();
    assert!(
        x.path == "//*"
            || x.path.ends_with("*/")
            || x.path.ends_with('*')
    );
}

#[test]
fn test_xpath_expr_join() {
    let x1 = XPathExpr::new("//", "a", "foo=1");
    let x2 = XPathExpr::new("/*/", "span", "bar=2");
    let r = x1.join("|", &x2, "::", true);
    // Type and content checks
    assert!(r.element.starts_with("span"));
    assert!(r.path.contains("|") || r.path.contains("::"));
}